### PR TITLE
Cython cleanup

### PR DESCRIPTION
**Pull request by LSchueler**
I cleaned up some of the older Cython code and I could actually reuse two functions and delete duplicate code.

### CODE REVIEW

File: gstools/field/summator.pyx [hunk 1]
```diff
@@ -12,31 +12,26 @@ from libc.math cimport sin, cos
...
-    cdef int dim
-    dim = pos.shape[0]
+    cdef int dim = pos.shape[0]
 
-    X_len = pos.shape[1]
-    N = cov_samples.shape[1]
+    cdef int X_len = pos.shape[1]
+    cdef int N = cov_samples.shape[1]
 
-    cdef double[:] summed_modes = np.zeros(X_len, dtype=DTYPE)
```
Comment (MuellerSeb) [inline gstools/field/summator.pyx:33]: So is using `dtype` really redundant here? It seems I know to little about cython-numpy integration.

File: gstools/field/summator.pyx [hunk 2]
```diff
@@ -12,31 +12,26 @@ from libc.math cimport sin, cos
...
+    cdef double[:] summed_modes = np.zeros(X_len)
 
     for i in prange(X_len, nogil=True):
         for j in range(N):
             phase = 0.
             for d in range(dim):
-                phase += cov_samples[d,j] * pos[d,i]
+                phase += cov_samples[d, j] * pos[d, i]
```
Comment (MuellerSeb) [inline gstools/field/summator.pyx:34]: Unfortunately `black` is still not able to deal with `*.pyx` files: https://github.com/psf/black/issues/359

File: gstools/variogram/estimator.pyx [hunk 3]
```diff
@@ -106,21 +105,21 @@ cdef inline double estimator_cressie(const double f_diff) nogil:
 ctypedef double (*_estimator_func)(const double) nogil
 
 cdef inline void normalization_matheron(
-    vector[double]& variogram,
-    vector[long]& counts
+    double[:] variogram,
+    long[:] counts,
```
Comment (MuellerSeb) [inline gstools/variogram/estimator.pyx:109]: I thought `vector` is faster? And is this still using the references? (formerly marked with `&`) Or is this done automatically by inlining the function? Again: I could use some training in cython.

File: gstools/variogram/estimator.pyx [hunk 4]
```diff
@@ -129,38 +128,30 @@ cdef inline void normalization_cressie(
...
     cdef long cnt
     for d in range(variogram.shape[0]):
-        for i in range(variogram.shape[1]):
-            # avoid division by zero
-            cnt = max(counts[d, i], 1)
-            variogram[d, i] = (
-                0.5 * (1./cnt * variogram[d, i])**4 /
-                (0.457 + 0.494 / cnt + 0.045 / cnt**2)
-            )
+        normalization_cressie(variogram[d, :], counts[d, :])
```
Comment (MuellerSeb) [inline gstools/variogram/estimator.pyx:150]: Good catch!

File: gstools/variogram/estimator.pyx [hunk 5]
```diff
@@ -129,38 +128,30 @@ cdef inline void normalization_cressie(
...
-    long[:,:]& counts
+    double[:,:] variogram,
+    long[:,:] counts,
 ):
     cdef int d, i
     for d in range(variogram.shape[0]):
-        for i in range(variogram.shape[1]):
-            # avoid division by zero
-            variogram[d, i] /= (2. * max(counts[d, i], 1))
+        normalization_matheron(variogram[d, :], counts[d, :])
```
Comment (MuellerSeb) [inline gstools/variogram/estimator.pyx:141]: Good catch!

File: gstools/variogram/estimator.pyx [hunk 6]
```diff
@@ -345,8 +332,8 @@ def ma_structured(
     cdef int j_max = f.shape[1]
     cdef int k_max = i_max + 1
 
-    cdef vector[double] variogram = vector[double](k_max, 0.0)
-    cdef vector[long] counts = vector[long](k_max, 0)
+    cdef double[:] variogram = np.zeros(k_max)
+    cdef long[:] counts = np.zeros(k_max, dtype=long)
```
Comment (MuellerSeb) [inline gstools/variogram/estimator.pyx:336]: Ah! So you need to specify dtype if it's not `double` right? I hope numpy doesn't change its default behavior to soon.

**Review by MuellerSeb — COMMENTED** [review]
Nice cleanup! One thing following "explicit is better than implicit": Wouldn't it be better to always use `dtype` when initializing numpy arrays in cython to not run into possible errors in the future? Maybe I am a bit overcautious here...

File: gstools/variogram/estimator.pyx (hunk 3 above)
Comment (LSchueler) [reply gstools/variogram/estimator.pyx:109]: It is faster, but I think only during the initialization of the variable. It does not call any Python code, in contrast to the `cdef double[:, :] variogram = np.zeros(...)`. But I'm very sure that that is negligible and I think the code is a lot cleaner when only using Cython's memoryviews and not mixed up with C++ data types.

**Comment by LSchueler** [comment]
I was actually also very unsure if the `dtype` stuff should go or not. Sure, explicit is better than implicit, but if Python would stay true to that, it shouldn't allow default arguments.
And I cannot imagine numpy to change its default `dtype` from 64 bit floats to anything else. But as I said, I'm pretty unsure about removing that. If you think we should stick to Python's (sometimes hypocritical ;-) ) zen code, then I'd be happy to change the code again.

**Comment by LSchueler** [comment]
@MuellerSeb Have you made up your mind about the `dtype`?

I think I will save the topic of contiguous arrays for another PR.

File: gstools/variogram/estimator.pyx (hunk 3 above)
Comment (MuellerSeb) [reply gstools/variogram/estimator.pyx:109]: Nice.

**Comment by MuellerSeb** [comment]
@LSchueler : I'd like to see an explicit `dtype` setting in the `np.zeros` call. And one other thing that came to my mind: Couldn't we remove the line `# distutils: language = c++` in `extimator.pyx` now since we don't use `vector`? After that I am happy to merge.

**Comment by MuellerSeb** [comment]
Also this line can be removed I guess:
https://github.com/GeoStat-Framework/GSTools/blob/3e4b6c36bd89fe3a8ca11fbde60f67fb1be0a5d2/setup.py#L146

**Comment by LSchueler** [comment]
I think we should keep the C++ flag. It doesn't hurt us, the same compiler is used anyway and this way, in case one day we want to use a C++ feature, we don't have to dig through the Cython docs again.

I added the dtypes again to the numpy functions.

**Pull request merged**